<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.552">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Kynon J Benjamin">
<meta name="dcterms.date" content="2024-07-03">

<title>Visualize local ancestry: chromosome section</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="01.visualization_files/libs/clipboard/clipboard.min.js"></script>
<script src="01.visualization_files/libs/quarto-html/quarto.js"></script>
<script src="01.visualization_files/libs/quarto-html/popper.min.js"></script>
<script src="01.visualization_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="01.visualization_files/libs/quarto-html/anchor.min.js"></script>
<link href="01.visualization_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="01.visualization_files/libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="01.visualization_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="01.visualization_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="01.visualization_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="dark">
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#load-packages" id="toc-load-packages" class="nav-link active" data-scroll-target="#load-packages"><span class="header-section-number">1</span> Load packages</a></li>
  <li><a href="#load-data" id="toc-load-data" class="nav-link" data-scroll-target="#load-data"><span class="header-section-number">2</span> Load data</a></li>
  <li><a href="#visualize-chromosome-global-ancestry" id="toc-visualize-chromosome-global-ancestry" class="nav-link" data-scroll-target="#visualize-chromosome-global-ancestry"><span class="header-section-number">3</span> Visualize chromosome global ancestry</a>
  <ul class="collapse">
  <li><a href="#organize-chromosome" id="toc-organize-chromosome" class="nav-link" data-scroll-target="#organize-chromosome"><span class="header-section-number">3.1</span> Organize chromosome</a></li>
  <li><a href="#create-and-save-the-plot" id="toc-create-and-save-the-plot" class="nav-link" data-scroll-target="#create-and-save-the-plot"><span class="header-section-number">3.2</span> Create and save the plot</a></li>
  <li><a href="#mean-and-median-african-ancestry-proportion" id="toc-mean-and-median-african-ancestry-proportion" class="nav-link" data-scroll-target="#mean-and-median-african-ancestry-proportion"><span class="header-section-number">3.3</span> Mean and median African ancestry proportion</a></li>
  </ul></li>
  <li><a href="#prepare-data" id="toc-prepare-data" class="nav-link" data-scroll-target="#prepare-data"><span class="header-section-number">4</span> Prepare data</a>
  <ul class="collapse">
  <li><a href="#helper-functions" id="toc-helper-functions" class="nav-link" data-scroll-target="#helper-functions"><span class="header-section-number">4.1</span> Helper functions</a></li>
  <li><a href="#define-column-names" id="toc-define-column-names" class="nav-link" data-scroll-target="#define-column-names"><span class="header-section-number">4.2</span> Define column names</a></li>
  <li><a href="#convert-data-to-dask-dataframe" id="toc-convert-data-to-dask-dataframe" class="nav-link" data-scroll-target="#convert-data-to-dask-dataframe"><span class="header-section-number">4.3</span> Convert data to dask dataframe</a></li>
  <li><a href="#combine-loci-with-haplotype-data" id="toc-combine-loci-with-haplotype-data" class="nav-link" data-scroll-target="#combine-loci-with-haplotype-data"><span class="header-section-number">4.4</span> Combine loci with haplotype data</a></li>
  <li><a href="#select-chromosomes" id="toc-select-chromosomes" class="nav-link" data-scroll-target="#select-chromosomes"><span class="header-section-number">4.5</span> Select chromosomes</a></li>
  </ul></li>
  <li><a href="#plot-a-section-of-each-chromosome" id="toc-plot-a-section-of-each-chromosome" class="nav-link" data-scroll-target="#plot-a-section-of-each-chromosome"><span class="header-section-number">5</span> Plot a section of each chromosome</a>
  <ul class="collapse">
  <li><a href="#helper-function" id="toc-helper-function" class="nav-link" data-scroll-target="#helper-function"><span class="header-section-number">5.1</span> Helper function</a></li>
  <li><a href="#seed-for-reproducibility" id="toc-seed-for-reproducibility" class="nav-link" data-scroll-target="#seed-for-reproducibility"><span class="header-section-number">5.2</span> Seed for reproducibility</a></li>
  <li><a href="#plotting" id="toc-plotting" class="nav-link" data-scroll-target="#plotting"><span class="header-section-number">5.3</span> Plotting</a></li>
  </ul></li>
  <li><a href="#session-information" id="toc-session-information" class="nav-link" data-scroll-target="#session-information"><span class="header-section-number">6</span> Session information</a></li>
  </ul>
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="visualization.ipynb" download="visualization.ipynb"><i class="bi bi-journal-code"></i>Jupyter</a></li></ul></div></nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Visualize local ancestry: chromosome section</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Kynon J Benjamin </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">July 3, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>

<nav id="TOC-body" role="doc-toc">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#load-packages" id="toc-load-packages"><span class="header-section-number">1</span> Load packages</a></li>
  <li><a href="#load-data" id="toc-load-data"><span class="header-section-number">2</span> Load data</a></li>
  <li><a href="#visualize-chromosome-global-ancestry" id="toc-visualize-chromosome-global-ancestry"><span class="header-section-number">3</span> Visualize chromosome global ancestry</a>
  <ul>
  <li><a href="#organize-chromosome" id="toc-organize-chromosome"><span class="header-section-number">3.1</span> Organize chromosome</a></li>
  <li><a href="#create-and-save-the-plot" id="toc-create-and-save-the-plot"><span class="header-section-number">3.2</span> Create and save the plot</a></li>
  <li><a href="#mean-and-median-african-ancestry-proportion" id="toc-mean-and-median-african-ancestry-proportion"><span class="header-section-number">3.3</span> Mean and median African ancestry proportion</a></li>
  </ul></li>
  <li><a href="#prepare-data" id="toc-prepare-data"><span class="header-section-number">4</span> Prepare data</a>
  <ul>
  <li><a href="#helper-functions" id="toc-helper-functions"><span class="header-section-number">4.1</span> Helper functions</a></li>
  <li><a href="#define-column-names" id="toc-define-column-names"><span class="header-section-number">4.2</span> Define column names</a></li>
  <li><a href="#convert-data-to-dask-dataframe" id="toc-convert-data-to-dask-dataframe"><span class="header-section-number">4.3</span> Convert data to dask dataframe</a></li>
  <li><a href="#combine-loci-with-haplotype-data" id="toc-combine-loci-with-haplotype-data"><span class="header-section-number">4.4</span> Combine loci with haplotype data</a></li>
  <li><a href="#select-chromosomes" id="toc-select-chromosomes"><span class="header-section-number">4.5</span> Select chromosomes</a></li>
  </ul></li>
  <li><a href="#plot-a-section-of-each-chromosome" id="toc-plot-a-section-of-each-chromosome"><span class="header-section-number">5</span> Plot a section of each chromosome</a>
  <ul>
  <li><a href="#helper-function" id="toc-helper-function"><span class="header-section-number">5.1</span> Helper function</a></li>
  <li><a href="#seed-for-reproducibility" id="toc-seed-for-reproducibility"><span class="header-section-number">5.2</span> Seed for reproducibility</a></li>
  <li><a href="#plotting" id="toc-plotting"><span class="header-section-number">5.3</span> Plotting</a></li>
  </ul></li>
  <li><a href="#session-information" id="toc-session-information"><span class="header-section-number">6</span> Session information</a></li>
  </ul>
</nav>
<section id="load-packages" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Load packages</h1>
<div id="d0bc069c-2341-4ef3-9e27-30fef3255c62" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> session_info</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pyhere <span class="im">import</span> here</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> rfmix_reader <span class="im">import</span> read_rfmix</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="load-data" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Load data</h1>
<div id="b97d453d-94ca-4015-af2c-81e01ef7f5f7" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>prefix_path <span class="op">=</span> here(<span class="st">"input/real_data/_m/"</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>binary_dir <span class="op">=</span> here(<span class="st">"real_data/gpu_version/_m/binary_files/"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>loci, rf_q, admix <span class="op">=</span> read_rfmix(prefix_path, binary_dir<span class="op">=</span>binary_dir)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GPU 0: NVIDIA A100 80GB PCIe
  Total memory: 79.15 GB
  CUDA capability: 8.0
Multiple files read in this order: ['chr1', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16', 'chr17', 'chr18', 'chr19', 'chr2', 'chr20', 'chr21', 'chr22', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9']</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Mapping loci files:   0%|          | 0/22 [00:00&lt;?, ?it/s]Mapping loci files:   5%|▍         | 1/22 [00:01&lt;00:29,  1.39s/it]Mapping loci files:   9%|▉         | 2/22 [00:01&lt;00:17,  1.17it/s]Mapping loci files:  14%|█▎        | 3/22 [00:02&lt;00:12,  1.48it/s]Mapping loci files:  18%|█▊        | 4/22 [00:02&lt;00:10,  1.73it/s]Mapping loci files:  23%|██▎       | 5/22 [00:03&lt;00:08,  2.03it/s]Mapping loci files:  27%|██▋       | 6/22 [00:03&lt;00:06,  2.35it/s]Mapping loci files:  32%|███▏      | 7/22 [00:03&lt;00:05,  2.65it/s]Mapping loci files:  36%|███▋      | 8/22 [00:03&lt;00:04,  2.85it/s]Mapping loci files:  41%|████      | 9/22 [00:04&lt;00:04,  3.04it/s]Mapping loci files:  45%|████▌     | 10/22 [00:04&lt;00:03,  3.05it/s]Mapping loci files:  50%|█████     | 11/22 [00:04&lt;00:03,  3.10it/s]Mapping loci files:  55%|█████▍    | 12/22 [00:05&lt;00:04,  2.15it/s]Mapping loci files:  59%|█████▉    | 13/22 [00:05&lt;00:03,  2.40it/s]Mapping loci files:  64%|██████▎   | 14/22 [00:06&lt;00:02,  2.95it/s]Mapping loci files:  68%|██████▊   | 15/22 [00:06&lt;00:01,  3.55it/s]Mapping loci files:  73%|███████▎  | 16/22 [00:06&lt;00:02,  2.62it/s]Mapping loci files:  77%|███████▋  | 17/22 [00:07&lt;00:02,  2.16it/s]Mapping loci files:  82%|████████▏ | 18/22 [00:08&lt;00:02,  1.99it/s]Mapping loci files:  86%|████████▋ | 19/22 [00:08&lt;00:01,  1.88it/s]Mapping loci files:  91%|█████████ | 20/22 [00:09&lt;00:01,  1.85it/s]Mapping loci files:  95%|█████████▌| 21/22 [00:10&lt;00:00,  1.63it/s]Mapping loci files: 100%|██████████| 22/22 [00:10&lt;00:00,  1.82it/s]Mapping loci files: 100%|██████████| 22/22 [00:10&lt;00:00,  2.10it/s]
Mapping Q files:   0%|          | 0/22 [00:00&lt;?, ?it/s]Mapping Q files:  14%|█▎        | 3/22 [00:00&lt;00:01, 12.17it/s]Mapping Q files: 100%|██████████| 22/22 [00:00&lt;00:00, 71.22it/s]
Mapping fb files:   0%|          | 0/22 [00:00&lt;?, ?it/s]Mapping fb files:   5%|▍         | 1/22 [00:00&lt;00:03,  6.33it/s]Mapping fb files:   9%|▉         | 2/22 [00:00&lt;00:02,  7.96it/s]Mapping fb files:  14%|█▎        | 3/22 [00:00&lt;00:02,  8.75it/s]Mapping fb files:  23%|██▎       | 5/22 [00:00&lt;00:01, 10.28it/s]Mapping fb files:  32%|███▏      | 7/22 [00:00&lt;00:01, 12.25it/s]Mapping fb files:  41%|████      | 9/22 [00:00&lt;00:00, 13.76it/s]Mapping fb files:  50%|█████     | 11/22 [00:01&lt;00:01, 10.73it/s]Mapping fb files:  59%|█████▉    | 13/22 [00:01&lt;00:00, 10.17it/s]Mapping fb files:  73%|███████▎  | 16/22 [00:01&lt;00:00, 11.84it/s]Mapping fb files:  82%|████████▏ | 18/22 [00:01&lt;00:00, 10.01it/s]Mapping fb files:  91%|█████████ | 20/22 [00:01&lt;00:00,  9.41it/s]Mapping fb files:  95%|█████████▌| 21/22 [00:02&lt;00:00,  9.26it/s]Mapping fb files: 100%|██████████| 22/22 [00:02&lt;00:00,  7.18it/s]Mapping fb files: 100%|██████████| 22/22 [00:02&lt;00:00,  9.36it/s]</code></pre>
</div>
</div>
</section>
<section id="visualize-chromosome-global-ancestry" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Visualize chromosome global ancestry</h1>
<section id="organize-chromosome" class="level2" data-number="3.1">
<h2 data-number="3.1" class="anchored" data-anchor-id="organize-chromosome"><span class="header-section-number">3.1</span> Organize chromosome</h2>
<div id="bf4eb07c-91e0-4c6f-8456-7ece16c7da00" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>chrom_order <span class="op">=</span> [<span class="ss">f'chr</span><span class="sc">{</span>i<span class="sc">}</span><span class="ss">'</span> <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>, <span class="dv">23</span>)]</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>rf_q_pandas <span class="op">=</span> rf_q.to_pandas() <span class="co"># Convert cuDF to pandas DataFrame</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>rf_q_pandas[<span class="st">'chrom'</span>] <span class="op">=</span> pd.Categorical(rf_q_pandas[<span class="st">'chrom'</span>], </span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>                                      categories<span class="op">=</span>chrom_order, ordered<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>rf_q_sorted <span class="op">=</span> rf_q_pandas.sort_values(<span class="st">'chrom'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="create-and-save-the-plot" class="level2" data-number="3.2">
<h2 data-number="3.2" class="anchored" data-anchor-id="create-and-save-the-plot"><span class="header-section-number">3.2</span> Create and save the plot</h2>
<div id="3aff334a-b851-4c6a-b3e3-56521ccaf7f0" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">15</span>, <span class="dv">8</span>))</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>sns.boxplot(x<span class="op">=</span><span class="st">'chrom'</span>, y<span class="op">=</span><span class="st">'AFR'</span>, data<span class="op">=</span>rf_q_sorted, </span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>            color<span class="op">=</span><span class="st">'lightgray'</span>, width<span class="op">=</span><span class="fl">0.6</span>)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>sns.stripplot(x<span class="op">=</span><span class="st">'chrom'</span>, y<span class="op">=</span><span class="st">'AFR'</span>, data<span class="op">=</span>rf_q_sorted, </span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>              color<span class="op">=</span><span class="st">'black'</span>, alpha<span class="op">=</span><span class="fl">0.1</span>, jitter<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>plt.axhline(y<span class="op">=</span><span class="fl">0.5</span>, color<span class="op">=</span><span class="st">'black'</span>, linestyle<span class="op">=</span><span class="st">'--'</span>, linewidth<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">'Global Ancestry (AFR) by Chromosome'</span>, fontsize<span class="op">=</span><span class="dv">18</span>)</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">'Chromosome'</span>, fontsize<span class="op">=</span><span class="dv">14</span>)</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">'African Genetic Ancestry'</span>, fontsize<span class="op">=</span><span class="dv">14</span>)</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>plt.xticks(rotation<span class="op">=</span><span class="dv">45</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>plt.yticks(fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="01.visualization_files/figure-html/cell-5-output-1.png" width="1430" height="758" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="mean-and-median-african-ancestry-proportion" class="level2" data-number="3.3">
<h2 data-number="3.3" class="anchored" data-anchor-id="mean-and-median-african-ancestry-proportion"><span class="header-section-number">3.3</span> Mean and median African ancestry proportion</h2>
<div id="4b0d7bc9-489e-4096-b4da-a210330fb790" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>rf_q_sorted.groupby(<span class="st">"chrom"</span>).agg({<span class="st">"AFR"</span>: [<span class="st">"mean"</span>, <span class="st">"median"</span>]})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="5">
<div>
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th colspan="2" data-quarto-table-cell-role="th" data-halign="left">AFR</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">mean</th>
<th data-quarto-table-cell-role="th">median</th>
</tr>
<tr class="header">
<th data-quarto-table-cell-role="th">chrom</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1</td>
<td>0.746126</td>
<td>0.768495</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr2</td>
<td>0.686543</td>
<td>0.713055</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr3</td>
<td>0.665584</td>
<td>0.690650</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr4</td>
<td>0.745466</td>
<td>0.769450</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr5</td>
<td>0.584738</td>
<td>0.608700</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr6</td>
<td>0.523042</td>
<td>0.548895</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr7</td>
<td>0.530944</td>
<td>0.558790</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr8</td>
<td>0.585027</td>
<td>0.609925</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr9</td>
<td>0.668068</td>
<td>0.686420</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr10</td>
<td>0.748637</td>
<td>0.781270</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr11</td>
<td>0.715288</td>
<td>0.741260</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr12</td>
<td>0.673626</td>
<td>0.701715</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr13</td>
<td>0.836891</td>
<td>0.871625</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr14</td>
<td>0.695394</td>
<td>0.723885</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr15</td>
<td>0.710502</td>
<td>0.746580</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr16</td>
<td>0.644983</td>
<td>0.675750</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr17</td>
<td>0.574651</td>
<td>0.603290</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr18</td>
<td>0.793579</td>
<td>0.828735</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr19</td>
<td>0.456887</td>
<td>0.491490</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr20</td>
<td>0.459954</td>
<td>0.500000</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr21</td>
<td>0.496484</td>
<td>0.542535</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr22</td>
<td>0.457181</td>
<td>0.500000</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</div>
<div id="d37d7c4a-8b89-4508-a0dc-4062bb109fc2" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(rf_q_sorted.AFR.mean(), rf_q_sorted.AFR.median())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>0.6363452039405462 0.6444000000000001</code></pre>
</div>
</div>
<p><strong>Select chromosome with:</strong> * High African Ancestry – chromosome 13 * Low African Ancestry – chromosome 19 * Average African Ancestry – chromosome 16</p>
</section>
</section>
<section id="prepare-data" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> Prepare data</h1>
<section id="helper-functions" class="level2" data-number="4.1">
<h2 data-number="4.1" class="anchored" data-anchor-id="helper-functions"><span class="header-section-number">4.1</span> Helper functions</h2>
<div id="1522b753-de35-4d7b-afe1-d23106884cac" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> dask.dataframe <span class="im">as</span> dd</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> multiprocessing <span class="im">import</span> cpu_count</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="a59e8ae0-3778-47be-b851-ae6b60f75edb" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="cf">try</span>:</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>    <span class="im">from</span> torch.cuda <span class="im">import</span> is_available</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="cf">except</span> <span class="pp">ModuleNotFoundError</span> <span class="im">as</span> e:</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">"Warning: PyTorch is not installed. Using CPU!"</span>)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> is_available():</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> <span class="va">False</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="75cb00bb-dbd5-4ab4-a335-67656ab6fca8" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> _get_pops(rf_q):</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> rf_q.drop([<span class="st">"sample_id"</span>, <span class="st">"chrom"</span>], axis<span class="op">=</span><span class="dv">1</span>).columns.values</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> _get_sample_names(rf_q):</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> is_available():</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> rf_q.sample_id.unique().to_arrow()</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> rf_q.sample_id.unique()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="define-column-names" class="level2" data-number="4.2">
<h2 data-number="4.2" class="anchored" data-anchor-id="define-column-names"><span class="header-section-number">4.2</span> Define column names</h2>
<div id="32edc7fd-a2ef-4d34-bb17-da619def6b23" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>pops <span class="op">=</span> _get_pops(rf_q)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>sample_ids <span class="op">=</span> _get_sample_names(rf_q)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="convert-data-to-dask-dataframe" class="level2" data-number="4.3">
<h2 data-number="4.3" class="anchored" data-anchor-id="convert-data-to-dask-dataframe"><span class="header-section-number">4.3</span> Convert data to dask dataframe</h2>
<div id="0d223fb2-ecfc-49b3-a585-c885c51c862b" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>parts <span class="op">=</span> cpu_count()</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>ncols <span class="op">=</span> admix.shape[<span class="dv">1</span>] <span class="op">//</span> <span class="bu">len</span>(pops)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>ddf <span class="op">=</span> dd.from_pandas(loci.to_pandas(), npartitions<span class="op">=</span>parts)</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>data_matrix <span class="op">=</span> admix[:, :ncols] <span class="co"># select the first pop only (pop2 is just 1-pop1)</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>dask_df <span class="op">=</span> dd.from_dask_array(data_matrix, columns<span class="op">=</span>sample_ids)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="combine-loci-with-haplotype-data" class="level2" data-number="4.4">
<h2 data-number="4.4" class="anchored" data-anchor-id="combine-loci-with-haplotype-data"><span class="header-section-number">4.4</span> Combine loci with haplotype data</h2>
<div id="44b83265-6fc7-48b7-9ec3-12d77bf46cb0" class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>ddf <span class="op">=</span> dd.concat([ddf, dask_df], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="kw">del</span> dask_df <span class="co"># remove for memory consumption</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="select-chromosomes" class="level2" data-number="4.5">
<h2 data-number="4.5" class="anchored" data-anchor-id="select-chromosomes"><span class="header-section-number">4.5</span> Select chromosomes</h2>
<div id="0c4170fa-5b76-4c93-aefe-c56a9725b70f" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>chrom13 <span class="op">=</span> ddf[ddf[<span class="st">"chromosome"</span>] <span class="op">==</span> <span class="st">"chr13"</span>]</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>chrom16 <span class="op">=</span> ddf[ddf[<span class="st">"chromosome"</span>] <span class="op">==</span> <span class="st">"chr16"</span>]</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>chrom19 <span class="op">=</span> ddf[ddf[<span class="st">"chromosome"</span>] <span class="op">==</span> <span class="st">"chr19"</span>]</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="kw">del</span> ddf</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="plot-a-section-of-each-chromosome" class="level1" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> Plot a section of each chromosome</h1>
<section id="helper-function" class="level2" data-number="5.1">
<h2 data-number="5.1" class="anchored" data-anchor-id="helper-function"><span class="header-section-number">5.1</span> Helper function</h2>
<div id="6281e611-5b48-4eff-be56-e66cedb0bf3c" class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> numpy <span class="im">import</span> random</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="a210960c-2b0a-400d-b63a-a2e5b236cfd7" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> select_random_section(df, section_size<span class="op">=</span><span class="dv">10000</span>):</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Get the minimum and maximum positions</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>    min_pos <span class="op">=</span> df[<span class="st">'physical_position'</span>].<span class="bu">min</span>().compute()</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>    max_pos <span class="op">=</span> df[<span class="st">'physical_position'</span>].<span class="bu">max</span>().compute()</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Randomly select a start position</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>    start_pos <span class="op">=</span> random.randint(min_pos, max_pos <span class="op">-</span> section_size)</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>    end_pos <span class="op">=</span> start_pos <span class="op">+</span> section_size</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Filter the DataFrame for the selected section</span></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>    section <span class="op">=</span> df[(df[<span class="st">'physical_position'</span>] <span class="op">&gt;=</span> start_pos) <span class="op">&amp;</span> (df[<span class="st">'physical_position'</span>] <span class="op">&lt;</span> end_pos)]</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>    <span class="co">##section = section.melt(id_vars=["chromosome", "physical_position", "i"], </span></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>    <span class="co">##                       var_name="BrNum", value_name="Haplotypes")</span></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> section.compute()  <span class="co"># Compute to bring data into memory</span></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_section(df, section_size, fname):</span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>    selected_section <span class="op">=</span> select_random_section(df, section_size)</span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>    selected_section <span class="op">=</span> selected_section.sort_values(<span class="st">"physical_position"</span>)</span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a>    plt.figure(figsize<span class="op">=</span>(<span class="dv">12</span>, <span class="dv">3</span>))</span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>    sns.scatterplot(data<span class="op">=</span>selected_section, x<span class="op">=</span><span class="st">'physical_position'</span>, y<span class="op">=</span><span class="st">'Br2585'</span>, </span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a>                    color<span class="op">=</span><span class="st">'black'</span>, legend<span class="op">=</span><span class="va">False</span>, s<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a>    plt.title(<span class="ss">f'Chromosome Section Plot (Positions </span><span class="sc">{</span>selected_section[<span class="st">"physical_position"</span>]<span class="sc">.</span><span class="bu">min</span>()<span class="sc">}</span><span class="ss"> to </span><span class="sc">{</span>selected_section[<span class="st">"physical_position"</span>]<span class="sc">.</span><span class="bu">max</span>()<span class="sc">}</span><span class="ss">)'</span>, fontsize<span class="op">=</span><span class="dv">16</span>)</span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a>    plt.xlabel(<span class="st">'Chromosome Position'</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb18-23"><a href="#cb18-23" aria-hidden="true" tabindex="-1"></a>    plt.ylabel(<span class="st">'Haplotypes'</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb18-24"><a href="#cb18-24" aria-hidden="true" tabindex="-1"></a>    plt.yticks([<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">2</span>])</span>
<span id="cb18-25"><a href="#cb18-25" aria-hidden="true" tabindex="-1"></a>    plt.grid(<span class="va">True</span>, linestyle<span class="op">=</span><span class="st">'--'</span>, alpha<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb18-26"><a href="#cb18-26" aria-hidden="true" tabindex="-1"></a>    plt.tight_layout()</span>
<span id="cb18-27"><a href="#cb18-27" aria-hidden="true" tabindex="-1"></a>    plt.savefig(<span class="ss">f'local_ancestry.</span><span class="sc">{</span>fname<span class="sc">}</span><span class="ss">.pdf'</span>, dpi<span class="op">=</span><span class="dv">300</span>, bbox_inches<span class="op">=</span><span class="st">'tight'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="seed-for-reproducibility" class="level2" data-number="5.2">
<h2 data-number="5.2" class="anchored" data-anchor-id="seed-for-reproducibility"><span class="header-section-number">5.2</span> Seed for reproducibility</h2>
<div id="c6d39ca2-a7df-4b78-9247-88683a05f82c" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>seed_value <span class="op">=</span> <span class="dv">13</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>random.seed(seed_value)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>section_size <span class="op">=</span> <span class="dv">1000000</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="plotting" class="level2" data-number="5.3">
<h2 data-number="5.3" class="anchored" data-anchor-id="plotting"><span class="header-section-number">5.3</span> Plotting</h2>
<div id="a5abd133-2ef4-420d-8419-46fc9718571a" class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>plot_section(chrom13, section_size, <span class="st">"chr13"</span>)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>plot_section(chrom16, section_size, <span class="st">"chr16"</span>)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>plot_section(chrom19, section_size, <span class="st">"chr19"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="01.visualization_files/figure-html/cell-18-output-1.png" width="1143" height="279" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="01.visualization_files/figure-html/cell-18-output-2.png" width="1143" height="279" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="01.visualization_files/figure-html/cell-18-output-3.png" width="1143" height="279" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="session-information" class="level1" data-number="6">
<h1 data-number="6"><span class="header-section-number">6</span> Session information</h1>
<div id="c07e9b12-28e4-4326-a341-b048e27f62e1" class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>session_info.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="18">
<details>
<summary>Click to view session information</summary>
<pre>-----
cudf                24.06.00
dask                2024.6.2
dask_expr           1.1.6
matplotlib          3.7.1
numpy               1.26.4
pandas              2.0.3
pyhere              1.0.0
rfmix_reader        0.1.14
seaborn             0.13.2
session_info        1.0.0
torch               2.3.1+cu121
-----
</pre>
<details>
<summary>Click to view modules imported as dependencies</summary>
<pre>PIL                         10.3.0
anyio                       NA
argcomplete                 NA
arrow                       1.2.3
asttokens                   NA
astunparse                  1.6.3
attr                        23.2.0
attrs                       23.2.0
babel                       2.10.3
backcall                    0.2.0
cachetools                  5.3.3
cairo                       1.20.1
certifi                     2024.02.02
cffi                        1.16.0
chardet                     4.0.0
charset_normalizer          3.3.2
click                       8.1.2
cloudpickle                 2.1.0
comm                        0.2.2
cuda                        12.4.0
cupy                        13.1.0
cupy_backends               NA
cupyx                       NA
cycler                      0.10.0
cython_runtime              NA
dateutil                    2.8.2
debugpy                     1.6.3
decorator                   5.1.1
defusedxml                  0.7.1
distributed                 2024.6.2
dot_parser                  NA
exceptiongroup              1.2.0
executing                   0.10.0
fastjsonschema              NA
fastrlock                   0.8.2
fontTools                   4.33.3
fqdn                        NA
fsspec                      2022.7.1
gi                          3.40.1
gio                         NA
glib                        NA
gobject                     NA
google                      NA
gtk                         NA
idna                        2.10
importlib_metadata          NA
importlib_resources         NA
iniconfig                   NA
ipykernel                   6.15.1
ipywidgets                  8.1.2
isoduration                 NA
jedi                        0.19.1
jinja2                      3.1.1
json5                       NA
jsonpointer                 2.3
jsonschema                  4.21.1
jsonschema_specifications   NA
jupyter_events              0.10.0
jupyter_server              2.13.0
jupyterlab_server           2.25.4
kaleido                     0.2.1
kiwisolver                  1.4.3
llvmlite                    0.42.0
locket                      NA
lxml                        4.9.1
markupsafe                  2.1.1
matplotlib_inline           0.1.6
mpl_toolkits                NA
msgpack                     1.0.4
nbformat                    5.4.0
netifaces                   0.10.6
numba                       0.59.0
numexpr                     2.10.1
nvtx                        NA
optree                      0.11.0
overrides                   NA
packaging                   21.3
parso                       0.8.3
patsy                       0.5.2
pexpect                     4.8.0
pickleshare                 0.7.5
pkg_resources               NA
platformdirs                3.8.0
plotly                      5.9.0
pluggy                      1.5.0
prometheus_client           NA
prompt_toolkit              3.0.30
psutil                      5.9.8
ptyprocess                  0.7.0
pure_eval                   0.2.2
py                          1.11.0
pyarrow                     16.1.0
pydev_ipython               NA
pydevconsole                NA
pydevd                      2.8.0
pydevd_file_utils           NA
pydevd_plugins              NA
pydevd_tracing              NA
pydot                       1.4.2
pygments                    2.18.0
pynvml                      11.4.1
pyparsing                   3.0.9
pytest                      8.2.1
pythonjsonlogger            NA
pytz                        2022.1
rapids_dask_dependency      NA
referencing                 NA
requests                    2.31.0
rfc3339_validator           0.1.4
rfc3986_validator           0.1.1
rmm                         24.06.00
rpds                        NA
scipy                       1.13.1
send2trash                  NA
setuptools                  70.0.0
six                         1.16.0
sniffio                     1.2.0
socks                       1.7.1
sortedcontainers            2.4.0
sphinxcontrib               NA
stack_data                  0.4.0
statsmodels                 0.14.0
tblib                       3.0.0
tenacity                    NA
tlz                         0.12.0
toolz                       0.12.0
torchgen                    NA
tornado                     6.2
tqdm                        4.66.4
traitlets                   5.9.0
typing_extensions           NA
uri_template                NA
urllib3                     1.26.18
wcwidth                     0.2.5
webcolors                   1.11.1
websocket                   1.4.0
yaml                        6.0
zict                        3.0.0
zipp                        NA
zmq                         25.0.0
zoneinfo                    NA
zstandard                   0.18.0
</pre>
</details> <!-- seems like this ends pre, so might as well be explicit -->
<pre>-----
IPython             8.4.0
jupyter_client      8.0.2
jupyter_core        5.2.0
jupyterlab          4.1.5
notebook            7.1.2
-----
Python 3.9.16 (main, Dec  8 2022, 00:00:00) [GCC 11.3.1 20221121 (Red Hat 11.3.1-4)]
Linux-5.14.0-284.11.1.el9_2.x86_64-x86_64-with-glibc2.34
-----
Session information updated at 2024-07-03 15:11
</pre>
</details>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>